### ISSUE TITLE
Event listeners should be lambda friendly

### ISSUE THREAD

**Issue by bradphelan** [markaren/threepp#236]
Currently it's not possible to pass lambdas as event listeners

	void EventDispatcher::addEventListener(const std::string& type, EventListener* listener) 

You enforce a virtual base type and also force to pass by pointer. Both of these make it impossible to use lambdas. This makes the code much more noisy forcing users to create classes instead of simple callbacks. Instead of a base class

```cpp
    struct EventListener {

        virtual void onEvent(Event& event) = 0;

        virtual ~EventListener() = default;
    };
```

It would be better simply to define

```cpp
    using EventListener = std::function<void(Event&event)>;
```
and pass and store the callback by value. The problem then becomes how to remove the event. Easy is to use a std::shared_ptr with and return it. So we end up with

```cpp
    std::shared_ptr<EventListenerHandle> addEventListener(std::function<void(Event&event)>);
```

The user then hangs onto the EventListenerHandle till they don't need it anymore. As soon as the last copy of this goes out of scope then the listener will be removed. I know you want to make this like threejs but c++ is different and you should take advantage of RAII lifetimes and lambdas. 

```cpp
appModel.listenerSubscription =  mesh.addEventListener(
    [](Event & event){ std::cout << "my mesh was clicked" << std::endl; }
);
```

**Comment by markaren** [comment]
Yeah, I totally agree that the EventListener API needs modifications (it was worse before). I'll take your suggestions into consideration.

**Comment by markaren** [comment]
While I like your idea with a handler, I hope ef1b5c18dfe133e62defd88b96099b7bd2e2771a would be a step in the right direction.

**Comment by markaren** [comment]
I think the main issue with the handle suggestion for this particular API is that internally one listener is re-used N times. I would then have to store the result in a container etc. So internally, the usage becomes messier.
